core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java

```diff
@@ -727,6 +724,163 @@ public void write(String s, byte[] bytes) {
+        ResultHandle fields = tc.invokeVirtualMethod(ofMethod(Class.class, "getDeclaredFields", Field[].class), clazz);
+
+        BranchResult notWeakResult = tc.ifFalse(registerClass.getMethodParam(1));
+        try (BytecodeCreator notWeakBranch = notWeakResult.trueBranch()) {
+            ResultHandle carray = notWeakBranch.newArray(Class.class, notWeakBranch.load(1));
+            notWeakBranch.writeArrayValue(carray, 0, clazz);
+            notWeakBranch.invokeStaticMethod(ofMethod(RUNTIME_REFLECTION, "register", void.class, Class[].class),
+                    carray);
+            try (BytecodeCreator constructorsBranch = notWeakBranch.ifTrue(registerClass.getMethodParam(2)).trueBranch()) {
+                constructorsBranch.invokeStaticMethod(
+                        ofMethod(RUNTIME_REFLECTION, "register", void.class, Executable[].class),
+                        constructors);
+            }
+            try (BytecodeCreator methodsBranch = notWeakBranch.ifTrue(registerClass.getMethodParam(3)).trueBranch()) {
+                methodsBranch.invokeStaticMethod(
+                        ofMethod(RUNTIME_REFLECTION, "register", void.class, Executable[].class),
+                        methods);
+            }
+            try (BytecodeCreator fieldsBranch = notWeakBranch.ifTrue(registerClass.getMethodParam(4)).trueBranch()) {
+                fieldsBranch.invokeStaticMethod(
+                        ofMethod(RUNTIME_REFLECTION, "register", void.class,
+                                boolean.class, boolean.class, Field[].class),
+                        registerClass.getMethodParam(5), registerClass.getMethodParam(6), fields);
+            }
+        }
+        try (BytecodeCreator weakBranch = notWeakResult.falseBranch()) {
+            weakBranch.invokeStaticMethod(WEAK_REFLECTION_REGISTRATION, registerClass.getMethodParam(7), clazz,
+                    registerClass.getMethodParam(2), registerClass.getMethodParam(3), registerClass.getMethodParam(4));
+        }
+        CatchBlockCreator cc = tc.addCatch(Throwable.class);
+        // cc.invokeVirtualMethod(ofMethod(Throwable.class, "printStackTrace", void.class), cc.getCaughtException());
+        registerClass.returnVoid();
+        return registerClass;
+    }
+
+    /**
+     * Adds the method {@code registerForReflection} that calls by reflection all methods whose name starts with
+     * {@code registerClass}.
+     * <p/>
+     * The generated code is equivalent to:
+     *
+     * <pre>{@code
+     * private static void registerForReflection(Feature.BeforeAnalysisAccess beforeAnalysisAccess) {
+     *     for (Method method : Feature.class.getDeclaredMethods()) {
+     *         if (method.getName().startsWith("registerClass")) {
+     *             try {
+     *                 method.invoke(null, beforeAnalysisAccess);
+     *             } catch (Exception e) {
+     *                 e.printStackTrace();
+     *             }
+     *         }
+     *     }
```

Comment:
No I don't, because a `registerClass*` method is now only created when one of `methodSet`, `fieldSet` or `ctorSet` is not empty or `serialization` is true to avoid having too many methods which causes a `ClassTooLargeException`

Reply:
Sure? My understanding is that you are generating a `registerClass*` method for every 100 entries.

Reply:
Just to be sure that we understand each other, is this the code that you have in mind?
```
    private static void registerForReflection(Feature.BeforeAnalysisAccess beforeAnalysisAccess, int maxSuffixNameIndex) {
        for (int i=0;i<=maxSuffixNameIndex;i++) {
            try {
                Method method = Feature.class.getDeclaredMethod("registerClasses" + i, BeforeAnalysisAccess.class);
                method.invoke(null, beforeAnalysisAccess);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }
```

Reply:
Yes, if you need to keep using reflection for the invocations. Otherwise you can generate `maxSuffixNameIndex` invocations and avoid the reflection as well. I would prefer the latter but if there is a good reason not to I am OK with the reflection as well.